OpenSim/Moco/MocoGoal/MocoImpulseTrackingGoal.cpp

```diff
@@ -0,0 +1,362 @@
+        m_denominator *= gravityAccelMagnitude;
+    }
+
+    // Get the ExternalLoads object.
+    std::unique_ptr<ExternalLoads> extLoadsFromFile;
+    const ExternalLoads* extLoads;
+    if (!getProperty_external_loads().empty()) {
+        OPENSIM_THROW_IF_FRMOBJ(!getProperty_external_loads_file().empty(),
+            Exception,
+            "Expected either an ExternalLoads file or object, but both "
+            "were provided.");
+        extLoads = &get_external_loads();
+    }
+    else if (!get_external_loads_file().empty()) {
+        extLoadsFromFile = OpenSim::make_unique<ExternalLoads>(
+            get_external_loads_file(), true);
+        extLoads = extLoadsFromFile.get();
+    }
+    else {
+        OPENSIM_THROW_FRMOBJ(Exception, "No ExternalLoads provided.");
+    }
+
+    // Spline the data.
+    const std::string dataFilePath =
+        convertRelativeFilePathToAbsoluteFromXMLDocument(
+            extLoads->getDocumentFileName(),
+            extLoads->getDataFileName());
+    TimeSeriesTable data(dataFilePath);
+    GCVSplineSet allRefSplines(data);
+
+    // Each ExternalForce has an applied_to_body property. For the ExternalForce
+    // to be properly paired with a group of contact force components, the
+    // contact force components must also apply forces to the same body. Here,
+    // we find which of the two bodies in each contact force component matches
+    // the ExternalForce body.
+    const auto& scaleFactors = getModel().getComponentList<MocoScaleFactor>();
+    for (int ig = 0; ig < getProperty_contact_groups().size(); ++ig) {
+        const auto& group = get_contact_groups(ig);
+
+        OPENSIM_THROW_IF_FRMOBJ(
+            !extLoads->contains(group.get_external_force_name()), Exception,
+            "External force '{}' not found.",
+            group.get_external_force_name());
+        const auto& extForce =
+            extLoads->get(group.get_external_force_name());
+
+        GroupInfo groupInfo;
+        for (int ic = 0; ic < group.getProperty_contact_force_paths().size();
+            ++ic) {
+            const auto& path = group.get_contact_force_paths(ic);
+            const auto& contactForce =
+                model.getComponent<SmoothSphereHalfSpaceForce>(path);
+
+            int recordOffset = findRecordOffset(group, contactForce,
+                extForce.get_applied_to_body());
+
+            groupInfo.contacts.push_back(
+                std::make_pair(&contactForce, recordOffset));
+        }
+
+        // Gather the relevant data splines for this contact group.
+        // We assume that the "x", "y", and "z" columns could have been in any
+        // order.
+        const std::string& forceID = extForce.get_force_identifier();
+        groupInfo.refSplines.cloneAndAppend(allRefSplines.get(forceID + "x"));
+        groupInfo.refSplines.cloneAndAppend(allRefSplines.get(forceID + "y"));
+        groupInfo.refSplines.cloneAndAppend(allRefSplines.get(forceID + "z"));
```

Comment:
Per the header file comment, we'll now only store one spline for the desired impulse axis. You'll need to get the correct label (i.e., forceID + "dir") based on the impulse axis, and then store the spline to `groupInfo.refSpline`.

Reply:
Update the comment too.